sds_data_manager/lambda_images/instruments/batch_starter.py

```diff
@@ -0,0 +1,425 @@
+            min(process_dates),
+            max(process_dates),
+        )
+
+        cur.execute(query, params)
+        column_names = [desc[0] for desc in cur.description]
+        records = cur.fetchall()
+
+        # Map the column names to the records
+        records_dicts = [dict(zip(column_names, record)) for record in records]
+
+        all_records.extend(records_dicts)
+
+    return all_records
+
+
+def remove_ingested(records, dependents_level, process_dates):
+    """
+    Identifies and returns a list of dependent instruments
+    that have not been ingested for the specified dates.
+
+    Parameters
+    ----------
+    records : list of dict
+        A list of dictionaries where each dictionary
+        corresponds to a record from the database.
+    dependents_level : list of dict
+        A list containing dictionaries of dependent
+        instruments and their levels.
+    process_dates : list
+        A list of date strings representing the dates
+        to be checked for missing ingestions.
+
+    Returns
+    -------
+    output : list of dict
+        A list of dictionaries where each dictionary
+        indicates a dependent instrument and its level
+        that has not been ingested for a given date.
+
+    """
+
+    output = []
+
+    records_set = {
+        (rec["date"].date(), rec["instrument"], rec["level"]) for rec in records
+    }
+
+    for dependent in dependents_level:
+        for date_str in process_dates:
+            date_obj = datetime.strptime(date_str, "%Y-%m-%d").date()
+
+            if (
+                date_obj,
+                dependent["instrument"].lower(),
+                dependent["level"].lower(),
+            ) not in records_set:
+                output.append(
+                    {
+                        "instrument": dependent["instrument"],
+                        "level": dependent["level"],
+                        "date": date_str,
+                    }
+                )
+
+    return output
+
+
+def query_dependencies(cur, uningested, version):
+    """
+    Queries and checks dependencies for a list of grouped records.
+
+    Parameters
+    ----------
+    cur : database cursor
+        The cursor object associated with the database connection.
+    uningested : list of dict
+        A list of dictionaries where each dictionary corresponds to a record
+        from the database with keys 'instrument', 'level', and 'date'.
+    version : int or str
+        The version number to be used when querying dependent records.
+
+    Returns
+    -------
+    instruments_to_process : list of dict
+        A list of dictionaries. Each dictionary corresponds to a record
+        for which all dependencies are met.
+
+    """
+
+    dir_path = os.path.dirname(os.path.realpath(__file__))
+    json_path = os.path.join(dir_path, "dependents.json")
+
+    with open(json_path) as f:
+        data = json.load(f)
+
+    instruments_to_process = []
+
+    for record in uningested:
```

Comment:
uningested part is little confusing but I think it will go away with upcoming changes.